NBitcoin/Protocol/NetworkAddress.cs

```diff
@@ -5,113 +5,421 @@
+			}	
 
-		public void Adjust()
-		{
-			var nNow = Utils.DateTimeToUnixTime(DateTimeOffset.UtcNow);
-			if (ntime <= 100000000 || ntime > nNow + 10 * 60)
-				ntime = nNow - 5 * 24 * 60 * 60;
+			addr = ipv6.Skip(skip).ToArray();
 		}
 
-		public IPEndPoint Endpoint
+		private bool SetTor(string address)
 		{
-			get
+			if (!address.EndsWith(".onion"))
 			{
-				return new IPEndPoint(new IPAddress(ip), port);
+				return false;
 			}
-			set
+
+			byte[] bytes;
+			try
+			{
+				bytes = Encoders.Base32.DecodeData(address.Substring(0, address.Length - ".onion".Length));
+			}
+			catch(FormatException)
 			{
-				port = (ushort)value.Port;
-				var ipBytes = value.Address.GetAddressBytes();
-				if (ipBytes.Length == 16)
+				return false;
+			}
+
+			if (bytes.Length == ADDR_TORV2_SIZE)
+			{
+				network = (byte)Network.Onion;
+				addr = bytes;
+				return true;
+			}
+			else if(bytes.Length == ADDR_TORV3_SIZE  + TORV3_ADDR_CHECKSUM_LEN + TORV3_ADDR_VERSION_LEN)
+			{
+				var version = bytes[bytes.Length - 1];
+
+				if (version != 3)
 				{
-					ip = ipBytes;
+					return false;
 				}
-				else if (ipBytes.Length == 4)
+
+				var pubkey = bytes.SafeSubarray( 0, ADDR_TORV3_SIZE);
+				var chksum = bytes.SafeSubarray(ADDR_TORV3_SIZE,  TORV3_ADDR_CHECKSUM_LEN);
+
+				var calculatedChecksum = CalculateChecksum(pubkey);
+
+				if (!Utils.ArrayEqual(chksum, calculatedChecksum) )
 				{
-					//Convert to ipv4 mapped to ipv6
-					//In these addresses, the first 80 bits are zero, the next 16 bits are one, and the remaining 32 bits are the IPv4 address
-					ip = new byte[16];
-					Array.Copy(ipBytes, 0, ip, 12, 4);
-					Array.Copy(new byte[] { 0xFF, 0xFF }, 0, ip, 10, 2);
+					return false;
 				}
-				else
-					throw new NotSupportedException("Invalid IP address type");
+
+				network = (byte)Network.Onion;
+				addr = pubkey;
+				return true;
+			}
+			return false;
+		}
+
+		public bool SetI2P(string address)
+		{
+			if (!address.EndsWith(".b32.i2p"))
+			{
+				return false;
+			}
+
+			var encoded = address.Substring(0, address.Length - ".b32.i2p".Length);
+			if (encoded.Length != 52)
+			{
+				return false;
+			}
+
+			byte[] bytes;
+			try
+			{
+				bytes = Encoders.Base32.DecodeData(encoded + "====");
+			}
+			catch(FormatException)
+			{
+				return false;
+			}
+
+			if (bytes.Length != ADDR_I2P_SIZE + 3)
+			{
+				return false;
+			}
+
+			network = (byte)Network.I2P;
+			addr = bytes.SafeSubarray(0, ADDR_I2P_SIZE);
+			return true;
+		}
+
+		public bool SetSpecial(string address)
+		{
+			return SetTor(address) || SetI2P(address);
+		}
+
+		private static byte[] CalculateChecksum(byte[] pubkey)
+		{
+			// TORv3 CHECKSUM = H(".onion checksum" | PUBKEY | VERSION)[:2]
+			var prefix = Encoding.UTF8.GetBytes(".onion checksum");
+
+			var hasher = new Sha3Digest(256);
+
+			hasher.BlockUpdate(prefix, 0, prefix.Length);
+			hasher.BlockUpdate(pubkey, 0, pubkey.Length);
+			hasher.BlockUpdate(new byte[]{3}, 0, 1);

```

Comment:
```suggestion
			hasher.Update((byte)3);
```

Maybe this can be a solution?

Reply:
Done.